tempesta_fw/t/functional/leaks/test_leak.py

```diff
@@ -0,0 +1,138 @@
+"""
+Testing for memory leaks
+"""
+
+__author__ = 'Tempesta Technologies, Inc.'
+__copyright__ = 'Copyright (C) 2017 Tempesta Technologies, Inc.'
+__license__ = 'GPL2'
+
+
+
+import re
+import StringIO
+from time import sleep
+
+from testers import stress
+from helpers import tf_cfg, control, tempesta
+
+class LeakTest(stress.StressTest):
+    """ Leaks testing """
+    config = 'cache 0;\n'
+    backend_connections = 10
+    memory_leak_thresold = 16*1024 # in kib
+
+    def drop_caches(self):
+        """ Drop caches """
+        self.tempesta.node.run_cmd("echo 3 > /proc/sys/vm/drop_caches")
+        sleep(1)
+
+    def file_exists(self, file):
+        """ Check existance of file on Tempesta host """
+        check_cmd = "if [ -e %s ]; then echo -n yes; fi" % file
+        [stdout, stderr] = self.tempesta.node.run_cmd(check_cmd)
+        if stdout != "yes":
+            return False
+        return True
+
+    def read_kmemleaks(self):
+        """ Get amount of kmemleak records """
+        kmemleakfile = "/sys/kernel/debug/kmemleak"
+        tf_cfg.dbg(2, "Reading kmemleak")
+        if not self.file_exists(kmemleakfile):
+            tf_cfg.dbg(1, "kmemleak file does not exists")
+            return -1
+        cmd = "cat %s | grep \"unreferenced object\" | wc -l" % kmemleakfile
+        [stdout, stderr] = self.tempesta.node.run_cmd(cmd)
+        return int(stdout)
+
+    def get_memory_line(self, name):
+        """ Get value from /proc/meminfo """
+        [stdout, stderr] = self.tempesta.node.run_cmd("cat /proc/meminfo")
+        line = re.search("%s:[ ]+([0-9]+)" % name, stdout)
+        if len(line.groups()) == 0:
+            return -1
+        return int(line.group(1))
+
+    def slab_memory(self):
+        """ Get amount of slab used memory """
+        self.drop_caches()
+        slabmem = self.get_memory_line("Slab")
+        return slabmem
+
+    def free_memory(self):
+        """ Measure total memory usage """
+        self.drop_caches()
+        freemem = self.get_memory_line("MemFree")
+        return freemem
+
+    def assert_tempesta(self):
+        """ We don't check that traffic is parsed correctly. Only detect leaks. """
+
+    def assert_clients(self):
+        """ Check only traffic size. We don't need to check responses. """
+        cl_req_cnt = 0
+        cl_conn_cnt = 0
+        for client in self.clients:
+            req, err = client.results()
+            cl_req_cnt += req
+            cl_conn_cnt += client.connections * self.pipelined_req
+        exp_min = cl_req_cnt
+        exp_max = cl_req_cnt + cl_conn_cnt
+        self.assertTrue(
+            self.tempesta.stats.cl_msg_received >= exp_min and
+            self.tempesta.stats.cl_msg_received <= exp_max
```

Comment:
Please provide assert message, it'll be easier to get, what is wrong with the test